# PANDAS: Introduction and Basics

In [25]:
import pandas as pd

In [26]:
#performing a reading operation of a file from a system
df=pd.read_csv('HH-Automobile.csv')

In [27]:
df

symboling  normalized_losses         make fuel_type aspiration  \
0            3                168  alfa-romero       gas        std   
1            3                168  alfa-romero       gas        std   
2            1                168  alfa-romero       gas        std   
3            2                164         audi       gas        std   
4            2                164         audi       gas        std   
..         ...                ...          ...       ...        ...   
196         -1                 95        volvo       gas        std   
197         -1                 95        volvo       gas      turbo   
198         -1                 95        volvo       gas        std   
199         -1                 95        volvo    diesel      turbo   
200         -1                 95        volvo       gas      turbo   

    number_of_doors   body_style drive_wheels engine_location  wheel_base  \
0               two  convertible          rwd           front        88.6   
1               two  convertible          rwd           front        88.6   
2               two    hatchback          rwd           front        94.5   
3              four        sedan          fwd           front        99.8   
4              four        sedan          4wd           front        99.4   
..              ...          ...          ...             ...         ...   
196            four        sedan          rwd           front       109.1   
197            four        sedan          rwd           front       109.1   
198            four        sedan          rwd           front       109.1   
199            four        sedan          rwd           front       109.1   
200            four        sedan          rwd           front       109.1   

     ...  engine_size  fuel_system  bore  stroke compression_ratio horsepower  \
0    ...          130         mpfi  3.47    2.68               9.0        111   
1    ...          130         mpfi  3.47    2.68               9.0        111   
2    ...          152         mpfi  2.68    3.47               9.0        154   
3    ...          109         mpfi  3.19    3.40              10.0        102   
4    ...          136         mpfi  3.19    3.40               8.0        115   
..   ...          ...          ...   ...     ...               ...        ...   
196  ...          141         mpfi  3.78    3.15               9.5        114   
197  ...          141         mpfi  3.78    3.15               8.7        160   
198  ...          173         mpfi  3.58    2.87               8.8        134   
199  ...          145          idi  3.01    3.40              23.0        106   
200  ...          141         mpfi  3.78    3.15               9.5        114   

     peak_rpm city_mpg  highway_mpg  price  
0        5000       21           27  13495  
1        5000       21           27  16500  
2        5000       19           26  16500  
3        5500       24           30  13950  
4        5500       18           22  17450  
..        ...      ...          ...    ...  
196      5400       23           28  16845  
197      5300       19           25  19045  
198      5500       18           23  21485  
199      4800       26           27  22470  
200      5400       19           25  22625  

[201 rows x 26 columns]

Here note row indexes are generated from 0, 1,2,3....to 201. It is to be noted that these indexes are not in the csv file itself.

In [28]:
df.head()

symboling  normalized_losses         make fuel_type aspiration  \
0          3                168  alfa-romero       gas        std   
1          3                168  alfa-romero       gas        std   
2          1                168  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   

  number_of_doors   body_style drive_wheels engine_location  wheel_base  ...  \
0             two  convertible          rwd           front        88.6  ...   
1             two  convertible          rwd           front        88.6  ...   
2             two    hatchback          rwd           front        94.5  ...   
3            four        sedan          fwd           front        99.8  ...   
4            four        sedan          4wd           front        99.4  ...   

   engine_size  fuel_system  bore  stroke compression_ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0        111   
1          130         mpfi  3.47    2.68               9.0        111   
2          152         mpfi  2.68    3.47               9.0        154   
3          109         mpfi  3.19    3.40              10.0        102   
4          136         mpfi  3.19    3.40               8.0        115   

   peak_rpm city_mpg  highway_mpg  price  
0      5000       21           27  13495  
1      5000       21           27  16500  
2      5000       19           26  16500  
3      5500       24           30  13950  
4      5500       18           22  17450  

[5 rows x 26 columns]

In [29]:
df.head(10)
#will give upto 9 indexes of record i.e 10 rows starting from first row i.e index 0

symboling  normalized_losses         make fuel_type aspiration  \
0          3                168  alfa-romero       gas        std   
1          3                168  alfa-romero       gas        std   
2          1                168  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   
5          2                161         audi       gas        std   
6          1                158         audi       gas        std   
7          1                168         audi       gas        std   
8          1                158         audi       gas      turbo   
9          2                192          bmw       gas        std   

  number_of_doors   body_style drive_wheels engine_location  wheel_base  ...  \
0             two  convertible          rwd           front        88.6  ...   
1             two  convertible          rwd           front        88.6  ...   
2             two    hatchback          rwd           front        94.5  ...   
3            four        sedan          fwd           front        99.8  ...   
4            four        sedan          4wd           front        99.4  ...   
5             two        sedan          fwd           front        99.8  ...   
6            four        sedan          fwd           front       105.8  ...   
7            four        wagon          fwd           front       105.8  ...   
8            four        sedan          fwd           front       105.8  ...   
9             two        sedan          rwd           front       101.2  ...   

   engine_size  fuel_system  bore  stroke compression_ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0        111   
1          130         mpfi  3.47    2.68               9.0        111   
2          152         mpfi  2.68    3.47               9.0        154   
3          109         mpfi  3.19    3.40              10.0        102   
4          136         mpfi  3.19    3.40               8.0        115   
5          136         mpfi  3.19    3.40               8.5        110   
6          136         mpfi  3.19    3.40               8.5        110   
7          136         mpfi  3.19    3.40               8.5        110   
8          131         mpfi  3.13    3.40               8.3        140   
9          108         mpfi  3.50    2.80               8.8        101   

   peak_rpm city_mpg  highway_mpg  price  
0      5000       21           27  13495  
1      5000       21           27  16500  
2      5000       19           26  16500  
3      5500       24           30  13950  
4      5500       18           22  17450  
5      5500       19           25  15250  
6      5500       19           25  17710  
7      5500       19           25  18920  
8      5500       17           20  23875  
9      5800       23           29  16430  

[10 rows x 26 columns]

In [30]:
df.tail(10)
#will give upto 9 indexes of record i.e 10 rows starting from n to n-10 row i.e index n-1 to n-11

symboling  normalized_losses   make fuel_type aspiration number_of_doors  \
191         -1                 74  volvo       gas        std            four   
192         -2                103  volvo       gas        std            four   
193         -1                 74  volvo       gas        std            four   
194         -2                103  volvo       gas      turbo            four   
195         -1                 74  volvo       gas      turbo            four   
196         -1                 95  volvo       gas        std            four   
197         -1                 95  volvo       gas      turbo            four   
198         -1                 95  volvo       gas        std            four   
199         -1                 95  volvo    diesel      turbo            four   
200         -1                 95  volvo       gas      turbo            four   

    body_style drive_wheels engine_location  wheel_base  ...  engine_size  \
191      wagon          rwd           front       104.3  ...          141   
192      sedan          rwd           front       104.3  ...          141   
193      wagon          rwd           front       104.3  ...          141   
194      sedan          rwd           front       104.3  ...          130   
195      wagon          rwd           front       104.3  ...          130   
196      sedan          rwd           front       109.1  ...          141   
197      sedan          rwd           front       109.1  ...          141   
198      sedan          rwd           front       109.1  ...          173   
199      sedan          rwd           front       109.1  ...          145   
200      sedan          rwd           front       109.1  ...          141   

     fuel_system  bore  stroke compression_ratio horsepower  peak_rpm  \
191         mpfi  3.78    3.15               9.5        114      5400   
192         mpfi  3.78    3.15               9.5        114      5400   
193         mpfi  3.78    3.15               9.5        114      5400   
194         mpfi  3.62    3.15               7.5        162      5100   
195         mpfi  3.62    3.15               7.5        162      5100   
196         mpfi  3.78    3.15               9.5        114      5400   
197         mpfi  3.78    3.15               8.7        160      5300   
198         mpfi  3.58    2.87               8.8        134      5500   
199          idi  3.01    3.40              23.0        106      4800   
200         mpfi  3.78    3.15               9.5        114      5400   

    city_mpg  highway_mpg  price  
191       23           28  13415  
192       24           28  15985  
193       24           28  16515  
194       17           22  18420  
195       17           22  18950  
196       23           28  16845  
197       19           25  19045  
198       18           23  21485  
199       26           27  22470  
200       19           25  22625  

[10 rows x 26 columns]

In [31]:
df= pd.read_csv('HH-Automobile.csv')

In [32]:
type(df)

pandas.core.frame.DataFrame

In [33]:
df.tail()

symboling  normalized_losses   make fuel_type aspiration number_of_doors  \
196         -1                 95  volvo       gas        std            four   
197         -1                 95  volvo       gas      turbo            four   
198         -1                 95  volvo       gas        std            four   
199         -1                 95  volvo    diesel      turbo            four   
200         -1                 95  volvo       gas      turbo            four   

    body_style drive_wheels engine_location  wheel_base  ...  engine_size  \
196      sedan          rwd           front       109.1  ...          141   
197      sedan          rwd           front       109.1  ...          141   
198      sedan          rwd           front       109.1  ...          173   
199      sedan          rwd           front       109.1  ...          145   
200      sedan          rwd           front       109.1  ...          141   

     fuel_system  bore  stroke compression_ratio horsepower  peak_rpm  \
196         mpfi  3.78    3.15               9.5        114      5400   
197         mpfi  3.78    3.15               8.7        160      5300   
198         mpfi  3.58    2.87               8.8        134      5500   
199          idi  3.01    3.40              23.0        106      4800   
200         mpfi  3.78    3.15               9.5        114      5400   

    city_mpg  highway_mpg  price  
196       23           28  16845  
197       19           25  19045  
198       18           23  21485  
199       26           27  22470  
200       19           25  22625  

[5 rows x 26 columns]

In [34]:
df1=df.to_csv("testpd.csv")

In [35]:
pwd()

'C:\\Users\\Hp\\Downloads\\DSM--iNeuron (Jupyter notebook)\\DATA SCIENCE MASTERS'

In [36]:
!ls

'ls' is not recognized as an internal or external command,
operable program or batch file.


In [37]:
pd.read_csv('testpd.csv')

Unnamed: 0  symboling  normalized_losses         make fuel_type  \
0             0          3                168  alfa-romero       gas   
1             1          3                168  alfa-romero       gas   
2             2          1                168  alfa-romero       gas   
3             3          2                164         audi       gas   
4             4          2                164         audi       gas   
..          ...        ...                ...          ...       ...   
196         196         -1                 95        volvo       gas   
197         197         -1                 95        volvo       gas   
198         198         -1                 95        volvo       gas   
199         199         -1                 95        volvo    diesel   
200         200         -1                 95        volvo       gas   

    aspiration number_of_doors   body_style drive_wheels engine_location  ...  \
0          std             two  convertible          rwd           front  ...   
1          std             two  convertible          rwd           front  ...   
2          std             two    hatchback          rwd           front  ...   
3          std            four        sedan          fwd           front  ...   
4          std            four        sedan          4wd           front  ...   
..         ...             ...          ...          ...             ...  ...   
196        std            four        sedan          rwd           front  ...   
197      turbo            four        sedan          rwd           front  ...   
198        std            four        sedan          rwd           front  ...   
199      turbo            four        sedan          rwd           front  ...   
200      turbo            four        sedan          rwd           front  ...   

     engine_size  fuel_system  bore  stroke  compression_ratio horsepower  \
0            130         mpfi  3.47    2.68                9.0        111   
1            130         mpfi  3.47    2.68                9.0        111   
2            152         mpfi  2.68    3.47                9.0        154   
3            109         mpfi  3.19    3.40               10.0        102   
4            136         mpfi  3.19    3.40                8.0        115   
..           ...          ...   ...     ...                ...        ...   
196          141         mpfi  3.78    3.15                9.5        114   
197          141         mpfi  3.78    3.15                8.7        160   
198          173         mpfi  3.58    2.87                8.8        134   
199          145          idi  3.01    3.40               23.0        106   
200          141         mpfi  3.78    3.15                9.5        114   

    peak_rpm  city_mpg highway_mpg  price  
0       5000        21          27  13495  
1       5000        21          27  16500  
2       5000        19          26  16500  
3       5500        24          30  13950  
4       5500        18          22  17450  
..       ...       ...         ...    ...  
196     5400        23          28  16845  
197     5300        19          25  19045  
198     5500        18          23  21485  
199     4800        26          27  22470  
200     5400        19          25  22625  

[201 rows x 27 columns]

Here it is to be noted that one extra column staring from index 0 is created.

In [39]:
#here we are trying to avoid certain column when it is not separated by a comma which is by default the symbol used to differentiate between columns
df=pd.read_csv('occupation.csv')

In [40]:
df

user_id|age|gender|occupation|zip_code
0                  1|24|M|technician|85711
1                       2|53|F|other|94043
2                      3|23|M|writer|32067
3                  4|24|M|technician|43537
4                       5|33|F|other|15213
..                                     ...
938                 939|26|F|student|33319
939           940|32|M|administrator|02215
940                 941|20|M|student|97229
941               942|48|F|librarian|78209
942                 943|22|M|student|77841

[943 rows x 1 columns]

now we are overwriting the current data by using separator.Shown as below

In [41]:
df=pd.read_csv('occupation.csv', sep='|')

In [42]:
df

user_id  age gender     occupation zip_code
0          1   24      M     technician    85711
1          2   53      F          other    94043
2          3   23      M         writer    32067
3          4   24      M     technician    43537
4          5   33      F          other    15213
..       ...  ...    ...            ...      ...
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

[943 rows x 5 columns]

By defualt the very first row is considered as a header. We can change it by using following syntax.

In [43]:
df=pd.read_csv('occupation.csv',sep='|', header=None)
df

0    1       2              3         4
0    user_id  age  gender     occupation  zip_code
1          1   24       M     technician     85711
2          2   53       F          other     94043
3          3   23       M         writer     32067
4          4   24       M     technician     43537
..       ...  ...     ...            ...       ...
939      939   26       F        student     33319
940      940   32       M  administrator     02215
941      941   20       M        student     97229
942      942   48       F      librarian     78209
943      943   22       M        student     77841

[944 rows x 5 columns]

In [44]:
#giving name of column by own
df=pd.read_csv('occupation.csv',sep='|', header=None,names=['a','b','c','d','e'])
df

a    b       c              d         e
0    user_id  age  gender     occupation  zip_code
1          1   24       M     technician     85711
2          2   53       F          other     94043
3          3   23       M         writer     32067
4          4   24       M     technician     43537
..       ...  ...     ...            ...       ...
939      939   26       F        student     33319
940      940   32       M  administrator     02215
941      941   20       M        student     97229
942      942   48       F      librarian     78209
943      943   22       M        student     77841

[944 rows x 5 columns]

In [45]:
#skipping some rows
df=pd.read_csv('occupation.csv',sep='|', header=None,names=['a','b','c','d','e'],skiprows=[0,2,4,5,6,7,8,9])
df

a   b  c              d      e
0      1  24  M     technician  85711
1      3  23  M         writer  32067
2     10  53  M         lawyer  90703
3     11  39  F          other  30329
4     12  28  F          other  06405
..   ...  .. ..            ...    ...
931  939  26  F        student  33319
932  940  32  M  administrator  02215
933  941  20  M        student  97229
934  942  48  F      librarian  78209
935  943  22  M        student  77841

[936 rows x 5 columns]

In [46]:
#na_values is NULL VALUES. i.e wherever we are able to find 1 it will consider it as null values
df=pd.read_csv('occupation.csv',sep='|',na_values=[1])
df
#these na_values can be a string as well also.

user_id  age gender     occupation zip_code
0        NaN   24      M     technician    85711
1        2.0   53      F          other    94043
2        3.0   23      M         writer    32067
3        4.0   24      M     technician    43537
4        5.0   33      F          other    15213
..       ...  ...    ...            ...      ...
938    939.0   26      F        student    33319
939    940.0   32      M  administrator    02215
940    941.0   20      M        student    97229
941    942.0   48      F      librarian    78209
942    943.0   22      M        student    77841

[943 rows x 5 columns]

In [48]:
df1=pd.read_csv('occupation.csv', sep='|')
df1

user_id  age gender     occupation zip_code
0          1   24      M     technician    85711
1          2   53      F          other    94043
2          3   23      M         writer    32067
3          4   24      M     technician    43537
4          5   33      F          other    15213
..       ...  ...    ...            ...      ...
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

[943 rows x 5 columns]

In [49]:
type(df1)

pandas.core.frame.DataFrame

In [50]:
#here it will give the statistics related to data
df.describe()

user_id         age
count  942.000000  943.000000
mean   472.500000   34.051962
std    272.076276   12.192740
min      2.000000    7.000000
25%    237.250000   25.000000
50%    472.500000   31.000000
75%    707.750000   43.000000
max    943.000000   73.000000

In [51]:
df.dtypes
#returns the type of datatypes of each column

user_id       float64
age             int64
gender         object
occupation     object
zip_code       object
dtype: object

describe():<br>
note that it returns insights only for some specific columns i.e only for numerical data by default <br>

After arraning the data in increasing order (ascending order) <br>

25% : data at the 25th percentile index (not the average of first 25% data) <br>
50% : data at the 50th percentile index <br>
75% : data at the 75th percentile index

In [52]:
#storing data to system

df.to_csv("resultsoccupation.csv")


In [54]:
#It is to be noted that the csv file saved here contains column of indexes 
#starting from 0,1,...which will appear when we read it again.
df.to_csv("resultsoccupation1.csv",index=False)

#now the file created will not 

In [56]:
df.to_csv("resultsoccupation2.csv",index=False, sep='#')

In [58]:
df.to_csv("resultsoccupation3.csv",index=False, sep='#', columns=['age','gender'])

In [59]:
df.tail()

user_id  age gender     occupation zip_code
938    939.0   26      F        student    33319
939    940.0   32      M  administrator    02215
940    941.0   20      M        student    97229
941    942.0   48      F      librarian    78209
942    943.0   22      M        student    77841

In [60]:
df.columns

Index(['user_id', 'age', 'gender', 'occupation', 'zip_code'], dtype='object')

In [61]:
df['age']

0      24
1      53
2      23
3      24
4      33
       ..
938    26
939    32
940    20
941    48
942    22
Name: age, Length: 943, dtype: int64

In [62]:
type(df['age'])

pandas.core.series.Series

Here it is to be noted that series is a data type which means single record, single column data.

In [63]:
df[['age']]
#here the column name is passed in form of list show it will show dataframe
#if is passed in form of simple word then it will in series form as in line[62]

age
0     24
1     53
2     23
3     24
4     33
..   ...
938   26
939   32
940   20
941   48
942   22

[943 rows x 1 columns]

In [64]:
type(df[['age']])

pandas.core.frame.DataFrame

In [66]:
df['age','occupation']
#will show error bcz by defualt it takes only one column
#for more than 1 column we need to pass in form of list

KeyError: ('age', 'occupation')

In [67]:
df[['age','occupation']]

age     occupation
0     24     technician
1     53          other
2     23         writer
3     24     technician
4     33          other
..   ...            ...
938   26        student
939   32  administrator
940   20        student
941   48      librarian
942   22        student

[943 rows x 2 columns]

In [69]:
#reading data from Excel sheet
pd.read_excel('Book3.xlsx')

omrrts              Full_Name  Gender            City       Country
0     10001       Cornelius Kujawa    Male           Tokyo         Japan
1     10002      Patrica Courville  Female  New York Metro           USA
2     10003          Sanford Xiong    Male       Sao Paulo        Brazil
3     10004           Allen Burrus    Male           Seoul   South Korea
4     10005      Kathrine Fritzler  Female     Mexico City        Mexico
..      ...                    ...     ...             ...           ...
146   10146         Bobby Greening    Male          Dallas           USA
147   10147         Johnathon Haug    Male          Boston           USA
148   10148            Etta Bosque  Female  Belo Horizonte        Brazil
149   10149        Tomas Coppinger    Male        Khartoum         Sudan
150   10150           Nanci Bonier  Female          Riyadh  Saudi Arabia

[151 rows x 5 columns]

In [70]:
#by defaulit is reading only sheet1, so in order to add other sheet we do as:
pd.read_excel('Book3.xlsx',sheet_name='Phone_Numbers')

omrrts Phone_Number
0     10001     555-1234
1     10002     555-1236
2     10003     555-1238
3     10004     555-1240
4     10005     555-1242
..      ...          ...
141   10146     555-1526
142   10147     555-1528
143   10148     555-1530
144   10149     555-1532
145   10150     555-1534

[146 rows x 2 columns]

Reading data from HTML page

In [71]:
df3= pd.read_html("https://www.basketball-reference.com/players/i/ibakase01.html?utm_source=direct&utm_medium=Share&utm_campaign=ShareTool")

In [72]:
df3

[  Unnamed: 0_level_0 Unnamed: 1_level_0 Unnamed: 2_level_0 Per 36 Minutes  \
               Season                Age                 Lg             FG   
 0            2022-23                 33                NBA            6.6   
 
                                   ...                                Shooting  \
     FGA   3P  3PA   FT  FTA  ORB  ...  AST  STL  BLK  TOV   PF   PTS      FG%   
 0  13.3  1.5  4.3  1.6  2.2  2.5  ...  2.3  0.5  1.4  1.9  3.7  16.3    0.496   
 
                Unnamed: 20_level_0  
     3P%    FT%               WS/48  
 0  0.36  0.751               0.125  
 
 [1 rows x 21 columns],
        Season   Age                                     Tm  \
 0     2008-09  19.0  Did Not Play (other pro league—Spain)   
 1     2009-10  20.0                                    OKC   
 2     2010-11  21.0                                    OKC   
 3     2011-12  22.0                                    OKC   
 4     2012-13  23.0                                    OKC  

In [73]:
type(df3)

list

In [75]:
#stores data in form of multiple tables form a html page

In [76]:
len(df3)
#@no of tables will be displayed

7

In [77]:
df3[0]

Unnamed: 0_level_0 Unnamed: 1_level_0 Unnamed: 2_level_0 Per 36 Minutes  \
              Season                Age                 Lg             FG   
0            2022-23                 33                NBA            6.6   

                                  ...                                Shooting  \
    FGA   3P  3PA   FT  FTA  ORB  ...  AST  STL  BLK  TOV   PF   PTS      FG%   
0  13.3  1.5  4.3  1.6  2.2  2.5  ...  2.3  0.5  1.4  1.9  3.7  16.3    0.496   

               Unnamed: 20_level_0  
    3P%    FT%               WS/48  
0  0.36  0.751               0.125  

[1 rows x 21 columns]

In [78]:
df3[3].to_csv("players_data.csv")

here the file will be stored from the 3rd table.

In [80]:
pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

converting json data into tabular format from a jSON file.

In [88]:
s = """{
"name" :"sawan",
"phone_no" : 45454545,
"comp":["ISM", "UPES" , "cognizant"]
}
"""

In [89]:
import json

In [91]:
s1=json.loads(s)

In [92]:
s1

{'name': 'sawan', 'phone_no': 45454545, 'comp': ['ISM', 'UPES', 'cognizant']}

In [93]:
#to cross check use json viewer online whether it is inputted in correct format or not

In [94]:
pd.DataFrame(s1)

name  phone_no       comp
0  sawan  45454545        ISM
1  sawan  45454545       UPES
2  sawan  45454545  cognizant

In [95]:
pd.DataFrame(s1,columns=['comp'])

comp
0        ISM
1       UPES
2  cognizant

In [96]:
import requests

In [97]:
res = requests.get('https://api.github.com/repos/pandas-dev/pandas/issues')

In [98]:
res

<Response [200]>

In [99]:
#converting this response into jason format
res.json()

[{'url': 'https://api.github.com/repos/pandas-dev/pandas/issues/48346',
  'repository_url': 'https://api.github.com/repos/pandas-dev/pandas',
  'labels_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/48346/labels{/name}',
  'comments_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/48346/comments',
  'events_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/48346/events',
  'html_url': 'https://github.com/pandas-dev/pandas/issues/48346',
  'id': 1358644556,
  'node_id': 'I_kwDOAA0YD85Q-0VM',
  'number': 48346,
  'title': 'Governance: decision making on technical topics',
  'user': {'login': 'datapythonista',
   'id': 10058240,
   'node_id': 'MDQ6VXNlcjEwMDU4MjQw',
   'avatar_url': 'https://avatars.githubusercontent.com/u/10058240?v=4',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/datapythonista',
   'html_url': 'https://github.com/datapythonista',
   'followers_url': 'https://api.github.com/users/datapythonista/followers',
   'follow

In [100]:
js=res.json()

In [101]:
#convet this data into tabular form & read titles 
#js[0] will give the very first data
#js[0] [title]  as it is a doctionary and data is vailable in key : value pair
js[0]['title']

'Governance: decision making on technical topics'

In [102]:
for i in range(len(js)):
    print(js[i]['title'])

Governance: decision making on technical topics
⬆️ UPGRADE: Autoupdate pre-commit config
BUG: `.describe()` on `unsigned` types is resulting in `object` type result
REF: de-duplicate take_nd
PERF: Performance improvement value_counts for masked arrays
BUG: MultiIndex.unique losing ea dtype
BUG: read_html(extract_links=all) with no header
 BUG: make Series handle dtype='int64' for string array
BUG: empty groupby iloc equals empty dataframe but not empty groupby dataframe
to_sql: Adding overwrite or ignore behavior for row conflicts
BUG: test for empty input construct
WEB: Update to `jupyterlite==0.1.0b12`
BUG: dataframe returned for alternative agg method call np.size and "size" are inconsistent
BUG: inconsistent behaviour for empty DataFrames
BUG: Summing a series of bools doesn't return a number
BUG: DataFrame.at allows adding new rows to a DataFrame
ENH: Expose ArrowIntervalType and ArrowPeriodType in the public API
dataframe.astype() exception message to include column name
BUG: Con

In [103]:
len(js)

30

In [104]:
#converting json into tabular form
pd.DataFrame(js,columns=['url','title'])

url  \
0   https://api.github.com/repos/pandas-dev/pandas...   
1   https://api.github.com/repos/pandas-dev/pandas...   
2   https://api.github.com/repos/pandas-dev/pandas...   
3   https://api.github.com/repos/pandas-dev/pandas...   
4   https://api.github.com/repos/pandas-dev/pandas...   
5   https://api.github.com/repos/pandas-dev/pandas...   
6   https://api.github.com/repos/pandas-dev/pandas...   
7   https://api.github.com/repos/pandas-dev/pandas...   
8   https://api.github.com/repos/pandas-dev/pandas...   
9   https://api.github.com/repos/pandas-dev/pandas...   
10  https://api.github.com/repos/pandas-dev/pandas...   
11  https://api.github.com/repos/pandas-dev/pandas...   
12  https://api.github.com/repos/pandas-dev/pandas...   
13  https://api.github.com/repos/pandas-dev/pandas...   
14  https://api.github.com/repos/pandas-dev/pandas...   
15  https://api.github.com/repos/pandas-dev/pandas...   
16  https://api.github.com/repos/pandas-dev/pandas...   
17  https://api.github.com/repos/pandas-dev/pandas...   
18  https://api.github.com/repos/pandas-dev/pandas...   
19  https://api.github.com/repos/pandas-dev/pandas...   
20  https://api.github.com/repos/pandas-dev/pandas...   
21  https://api.github.com/repos/pandas-dev/pandas...   
22  https://api.github.com/repos/pandas-dev/pandas...   
23  https://api.github.com/repos/pandas-dev/pandas...   
24  https://api.github.com/repos/pandas-dev/pandas...   
25  https://api.github.com/repos/pandas-dev/pandas...   
26  https://api.github.com/repos/pandas-dev/pandas...   
27  https://api.github.com/repos/pandas-dev/pandas...   
28  https://api.github.com/repos/pandas-dev/pandas...   
29  https://api.github.com/repos/pandas-dev/pandas...   

                                                title  
0     Governance: decision making on technical topics  
1            ⬆️ UPGRADE: Autoupdate pre-commit config  
2   BUG: `.describe()` on `unsigned` types is resu...  
3                           REF: de-duplicate take_nd  
4   PERF: Performance improvement value_counts for...  
5              BUG: MultiIndex.unique losing ea dtype  
6    BUG: read_html(extract_links=all) with no header  
7    BUG: make Series handle dtype='int64' for str...  
8   BUG: empty groupby iloc equals empty dataframe...  
9   to_sql: Adding overwrite or ignore behavior fo...  
10                BUG: test for empty input construct  
11             WEB: Update to `jupyterlite==0.1.0b12`  
12  BUG: dataframe returned for alternative agg me...  
13   BUG: inconsistent behaviour for empty DataFrames  
14  BUG: Summing a series of bools doesn't return ...  
15  BUG: DataFrame.at allows adding new rows to a ...  
16  ENH: Expose ArrowIntervalType and ArrowPeriodT...  
17  dataframe.astype() exception message to includ...  
18  BUG: Converting period index to datetime index...  
19  BUG: TypeError exception when using read_html ...  
20  BUG: timedelta64[s] series constructor isn't e...  
21    ENH: Add ignore_index to Series.drop_duplicates  
22  Select_dtypes considers numpy int and pandas n...  
23  The indexes of DataFrame.describe(percentiles=...  
24  BUG: DataFrame.pivot not respecting None as co...  
25  ENH: Allow to skip series type check in assert...  
26  CLN, STYLE: remove unused variables and import...  
27  ENH: Allow to skip series type check in assert...  
28                     WEB: Unpin pydata sphinx theme  
29               BLD: Build wheels using cibuildwheel

In [108]:
pd.DataFrame(js)

url  \
0   https://api.github.com/repos/pandas-dev/pandas...   
1   https://api.github.com/repos/pandas-dev/pandas...   
2   https://api.github.com/repos/pandas-dev/pandas...   
3   https://api.github.com/repos/pandas-dev/pandas...   
4   https://api.github.com/repos/pandas-dev/pandas...   
5   https://api.github.com/repos/pandas-dev/pandas...   
6   https://api.github.com/repos/pandas-dev/pandas...   
7   https://api.github.com/repos/pandas-dev/pandas...   
8   https://api.github.com/repos/pandas-dev/pandas...   
9   https://api.github.com/repos/pandas-dev/pandas...   
10  https://api.github.com/repos/pandas-dev/pandas...   
11  https://api.github.com/repos/pandas-dev/pandas...   
12  https://api.github.com/repos/pandas-dev/pandas...   
13  https://api.github.com/repos/pandas-dev/pandas...   
14  https://api.github.com/repos/pandas-dev/pandas...   
15  https://api.github.com/repos/pandas-dev/pandas...   
16  https://api.github.com/repos/pandas-dev/pandas...   
17  https://api.github.com/repos/pandas-dev/pandas...   
18  https://api.github.com/repos/pandas-dev/pandas...   
19  https://api.github.com/repos/pandas-dev/pandas...   
20  https://api.github.com/repos/pandas-dev/pandas...   
21  https://api.github.com/repos/pandas-dev/pandas...   
22  https://api.github.com/repos/pandas-dev/pandas...   
23  https://api.github.com/repos/pandas-dev/pandas...   
24  https://api.github.com/repos/pandas-dev/pandas...   
25  https://api.github.com/repos/pandas-dev/pandas...   
26  https://api.github.com/repos/pandas-dev/pandas...   
27  https://api.github.com/repos/pandas-dev/pandas...   
28  https://api.github.com/repos/pandas-dev/pandas...   
29  https://api.github.com/repos/pandas-dev/pandas...   

                                    repository_url  \
0   https://api.github.com/repos/pandas-dev/pandas   
1   https://api.github.com/repos/pandas-dev/pandas   
2   https://api.github.com/repos/pandas-dev/pandas   
3   https://api.github.com/repos/pandas-dev/pandas   
4   https://api.github.com/repos/pandas-dev/pandas   
5   https://api.github.com/repos/pandas-dev/pandas   
6   https://api.github.com/repos/pandas-dev/pandas   
7   https://api.github.com/repos/pandas-dev/pandas   
8   https://api.github.com/repos/pandas-dev/pandas   
9   https://api.github.com/repos/pandas-dev/pandas   
10  https://api.github.com/repos/pandas-dev/pandas   
11  https://api.github.com/repos/pandas-dev/pandas   
12  https://api.github.com/repos/pandas-dev/pandas   
13  https://api.github.com/repos/pandas-dev/pandas   
14  https://api.github.com/repos/pandas-dev/pandas   
15  https://api.github.com/repos/pandas-dev/pandas   
16  https://api.github.com/repos/pandas-dev/pandas   
17  https://api.github.com/repos/pandas-dev/pandas   
18  https://api.github.com/repos/pandas-dev/pandas   
19  https://api.github.com/repos/pandas-dev/pandas   
20  https://api.github.com/repos/pandas-dev/pandas   
21  https://api.github.com/repos/pandas-dev/pandas   
22  https://api.github.com/repos/pandas-dev/pandas   
23  https://api.github.com/repos/pandas-dev/pandas   
24  https://api.github.com/repos/pandas-dev/pandas   
25  https://api.github.com/repos/pandas-dev/pandas   
26  https://api.github.com/repos/pandas-dev/pandas   
27  https://api.github.com/repos/pandas-dev/pandas   
28  https://api.github.com/repos/pandas-dev/pandas   
29  https://api.github.com/repos/pandas-dev/pandas   

                                           labels_url  \
0   https://api.github.com/repos/pandas-dev/pandas...   
1   https://api.github.com/repos/pandas-dev/pandas...   
2   https://api.github.com/repos/pandas-dev/pandas...   
3   https://api.github.com/repos/pandas-dev/pandas...   
4   https://api.github.com/repos/pandas-dev/pandas...   
5   https://api.github.com/repos/pandas-dev/pandas...   
6   https://api.github.com/repos/pandas-dev/pandas...   
7   https://api.github.com/repos/pandas-dev/pandas...   
8   https://api.github.com/repos/pandas-dev/pandas...   
9   https://api.github.com/repos/pand

In [109]:
df4=pd.DataFrame(js)

In [110]:
df4.to_csv('json_dump_csv')

In [112]:
js[0]['user']

{'login': 'datapythonista',
 'id': 10058240,
 'node_id': 'MDQ6VXNlcjEwMDU4MjQw',
 'avatar_url': 'https://avatars.githubusercontent.com/u/10058240?v=4',
 'gravatar_id': '',
 'url': 'https://api.github.com/users/datapythonista',
 'html_url': 'https://github.com/datapythonista',
 'followers_url': 'https://api.github.com/users/datapythonista/followers',
 'following_url': 'https://api.github.com/users/datapythonista/following{/other_user}',
 'gists_url': 'https://api.github.com/users/datapythonista/gists{/gist_id}',
 'starred_url': 'https://api.github.com/users/datapythonista/starred{/owner}{/repo}',
 'subscriptions_url': 'https://api.github.com/users/datapythonista/subscriptions',
 'organizations_url': 'https://api.github.com/users/datapythonista/orgs',
 'repos_url': 'https://api.github.com/users/datapythonista/repos',
 'events_url': 'https://api.github.com/users/datapythonista/events{/privacy}',
 'received_events_url': 'https://api.github.com/users/datapythonista/received_events',
 'type'

In [114]:
#conversion of user data into DataFrame format for user
lis=[]
for i in range(len(js)):
    lis.append(js[i]['user'])
pd.DataFrame(lis)


login        id               node_id  \
0        datapythonista  10058240  MDQ6VXNlcjEwMDU4MjQw   
1   github-actions[bot]  41898282  MDM6Qm90NDE4OTgyODI=   
2         galipremsagar  11664259  MDQ6VXNlcjExNjY0MjU5   
3          jbrockmendel   8078968  MDQ6VXNlcjgwNzg5Njg=   
4                 phofl  61934744  MDQ6VXNlcjYxOTM0NzQ0   
5                 phofl  61934744  MDQ6VXNlcjYxOTM0NzQ0   
6             mroeschke  10647082  MDQ6VXNlcjEwNjQ3MDgy   
7                   5j9   1055747  MDQ6VXNlcjEwNTU3NDc=   
8              Th3nn3ss  25160953  MDQ6VXNlcjI1MTYwOTUz   
9           JustinGuese   9588843  MDQ6VXNlcjk1ODg4NDM=   
10             Th3nn3ss  25160953  MDQ6VXNlcjI1MTYwOTUz   
11                jtpio    591645  MDQ6VXNlcjU5MTY0NQ==   
12             Th3nn3ss  25160953  MDQ6VXNlcjI1MTYwOTUz   
13             Th3nn3ss  25160953  MDQ6VXNlcjI1MTYwOTUz   
14             marberts  62676717  MDQ6VXNlcjYyNjc2NzE3   
15           rhshadrach  45562402  MDQ6VXNlcjQ1NTYyNDAy   
16            mroeschke  10647082  MDQ6VXNlcjEwNjQ3MDgy   
17             dannyi96  15380498  MDQ6VXNlcjE1MzgwNDk4   
18         paulsbrookes  15247443  MDQ6VXNlcjE1MjQ3NDQz   
19             tysonite    196535  MDQ6VXNlcjE5NjUzNQ==   
20             Th3nn3ss  25160953  MDQ6VXNlcjI1MTYwOTUz   
21            mroeschke  10647082  MDQ6VXNlcjEwNjQ3MDgy   
22              weikhor  36619387  MDQ6VXNlcjM2NjE5Mzg3   
23              weikhor  36619387  MDQ6VXNlcjM2NjE5Mzg3   
24                phofl  61934744  MDQ6VXNlcjYxOTM0NzQ0   
25              abalkin    535197  MDQ6VXNlcjUzNTE5Nw==   
26         MarcoGorelli  33491632  MDQ6VXNlcjMzNDkxNjMy   
27              abalkin    535197  MDQ6VXNlcjUzNTE5Nw==   
28                phofl  61934744  MDQ6VXNlcjYxOTM0NzQ0   
29            lithomas1  47963215  MDQ6VXNlcjQ3OTYzMjE1   

                                           avatar_url gravatar_id  \
0   https://avatars.githubusercontent.com/u/100582...               
1   https://avatars.githubusercontent.com/in/15368...               
2   https://avatars.githubusercontent.com/u/116642...               
3   https://avatars.githubusercontent.com/u/807896...               
4   https://avatars.githubusercontent.com/u/619347...               
5   https://avatars.githubusercontent.com/u/619347...               
6   https://avatars.githubusercontent.com/u/106470...               
7   https://avatars.githubusercontent.com/u/105574...               
8   https://avatars.githubusercontent.com/u/251609...               
9   https://avatars.githubusercontent.com/u/958884...               
10  https://avatars.githubusercontent.com/u/251609...               
11  https://avatars.githubusercontent.com/u/591645...               
12  https://avatars.githubusercontent.com/u/251609...               
13  https://avatars.githubusercontent.com/u/251609...               
14  https://avatars.githubusercontent.com/u/626767...               
15  https://avatars.githubusercontent.com/u/455624...               
16  https://avatars.githubusercontent.com/u/106470...               
17  https://avatars.githubusercontent.com/u/153804...               
18  https://avatars.githubusercontent.com/u/152474...               
19  https://avatars.githubusercontent.com/u/196535...               
20  https://avatars.githubusercontent.com/u/251609...               
21  https://avatars.githubusercontent.com/u/106470...               
22  https://avatars.githubusercontent.com/u/366193...               
23  https://avatars.githubusercontent.com/u/366193...               
24  https://avatars.githubusercontent.com/u/619347...               
25  https://avatars.githubusercontent.com/u/535197...               
26  https://avatars.githubusercontent.com/u/334916...               
27  https://avatars.githubusercontent.com/u/535197...               
28  https://avatars.githubusercontent.com/u/619347...               
29  https://avatars.githubusercontent.com/u/479632...               

                                                

In [116]:
# 2nd approach
json_data = pd.read_json("https://api.github.com/repos/pandas-dev/pandas/issues")
data = pd.DataFrame.from_records(json_data['user'])  #FROM_RECORDS IS A INBUILT FUNCTION
data.head()


login        id               node_id  \
0       datapythonista  10058240  MDQ6VXNlcjEwMDU4MjQw   
1  github-actions[bot]  41898282  MDM6Qm90NDE4OTgyODI=   
2        galipremsagar  11664259  MDQ6VXNlcjExNjY0MjU5   
3         jbrockmendel   8078968  MDQ6VXNlcjgwNzg5Njg=   
4                phofl  61934744  MDQ6VXNlcjYxOTM0NzQ0   

                                          avatar_url gravatar_id  \
0  https://avatars.githubusercontent.com/u/100582...               
1  https://avatars.githubusercontent.com/in/15368...               
2  https://avatars.githubusercontent.com/u/116642...               
3  https://avatars.githubusercontent.com/u/807896...               
4  https://avatars.githubusercontent.com/u/619347...               

                                                 url  \
0        https://api.github.com/users/datapythonista   
1  https://api.github.com/users/github-actions%5B...   
2         https://api.github.com/users/galipremsagar   
3          https://api.github.com/users/jbrockmendel   
4                 https://api.github.com/users/phofl   

                                 html_url  \
0       https://github.com/datapythonista   
1  https://github.com/apps/github-actions   
2        https://github.com/galipremsagar   
3         https://github.com/jbrockmendel   
4                https://github.com/phofl   

                                       followers_url  \
0  https://api.github.com/users/datapythonista/fo...   
1  https://api.github.com/users/github-actions%5B...   
2  https://api.github.com/users/galipremsagar/fol...   
3  https://api.github.com/users/jbrockmendel/foll...   
4       https://api.github.com/users/phofl/followers   

                                       following_url  \
0  https://api.github.com/users/datapythonista/fo...   
1  https://api.github.com/users/github-actions%5B...   
2  https://api.github.com/users/galipremsagar/fol...   
3  https://api.github.com/users/jbrockmendel/foll...   
4  https://api.github.com/users/phofl/following{/...   

                                           gists_url  \
0  https://api.github.com/users/datapythonista/gi...   
1  https://api.github.com/users/github-actions%5B...   
2  https://api.github.com/users/galipremsagar/gis...   
3  https://api.github.com/users/jbrockmendel/gist...   
4  https://api.github.com/users/phofl/gists{/gist...   

                                         starred_url  \
0  https://api.github.com/users/datapythonista/st...   
1  https://api.github.com/users/github-actions%5B...   
2  https://api.github.com/users/galipremsagar/sta...   
3  https://api.github.com/users/jbrockmendel/star...   
4  https://api.github.com/users/phofl/starred{/ow...   

                                   subscriptions_url  \
0  https://api.github.com/users/datapythonista/su...   
1  https://api.github.com/users/github-actions%5B...   
2  https://api.github.com/users/galipremsagar/sub...   
3  https://api.github.com/users/jbrockmendel/subs...   
4   https://api.github.com/users/phofl/subscriptions   

                                   organizations_url  \
0   https://api.github.com/users/datapythonista/orgs   
1  https://api.github.com/users/github-actions%5B...   
2    https://api.github.com/users/galipremsagar/orgs   
3     https://api.github.com/users/jbrockmendel/orgs   
4            https://api.github.com/users/phofl/orgs   

                                           repos_url  \
0  https://api.github.com/users/datapythonista/repos   
1  https://api.github.com/users/github-actions%5B...   
2   https://api.github.com/users/galipremsagar/repos   
3    https://api.github.com/users/jbrockmendel/repos   
4           https://api.github.com/users/phofl/repos   

                                          events_url  \
0  https://api.github.com/users/datapythonista/ev...   
1  https://api.github.com/users/github-actions%5B...   
2  https://api.github.com/users/galipremsagar/eve...   
3  https://api.github.com/users/jbrockmendel/even...   
4  https://a

## BASICS OF PANDAS (contdd..)

In [1]:
import pandas as pd

In [3]:
form_file = pd.read_csv('players_data.csv')

In [4]:
#pickle format is used to convert into binary format, by serialising or de-serialising7
form_file.to_pickle("my_pickle")

In [5]:
pwd()

'C:\\Users\\Hp\\Downloads\\DSM--iNeuron (Jupyter notebook)\\DATA SCIENCE MASTERS'

File created by pickle allows objects to be serialized to files on disk and deserialized back into the program at runtime; saves a byte stream that represents the objects; more often uses the . P extension rather than ". pickle."
it wont be able to read it directly, based on requirement it will be used.

In [6]:
form_file

Unnamed: 0     Season   Age   Tm   Lg  Pos      G     GS       MP      FG  \
0            0    2009-10  20.0  OKC  NBA   PF   73.0    0.0   1323.0   201.0   
1            1    2010-11  21.0  OKC  NBA   PF   82.0   44.0   2216.0   335.0   
2            2    2011-12  22.0  OKC  NBA   PF   66.0   66.0   1792.0   262.0   
3            3    2012-13  23.0  OKC  NBA   PF   80.0   80.0   2486.0   446.0   
4            4    2013-14  24.0  OKC  NBA   PF   81.0   81.0   2666.0   524.0   
5            5    2014-15  25.0  OKC  NBA   PF   64.0   64.0   2116.0   374.0   
6            6    2015-16  26.0  OKC  NBA   PF   78.0   78.0   2500.0   416.0   
7            7    2016-17  27.0  TOT  NBA   PF   79.0   79.0   2422.0   471.0   
8            8    2016-17  27.0  ORL  NBA   PF   56.0   56.0   1710.0   343.0   
9            9    2016-17  27.0  TOR  NBA   PF   23.0   23.0    712.0   128.0   
10          10    2017-18  28.0  TOR  NBA   PF   76.0   76.0   2093.0   379.0   
11          11    2018-19  29.0  TOR  NBA    C   74.0   51.0   2010.0   464.0   
12          12    2019-20  30.0  TOR  NBA    C   55.0   27.0   1485.0   343.0   
13          13    2020-21  31.0  LAC  NBA    C   41.0   39.0    955.0   186.0   
14          14    2021-22  32.0  TOT  NBA    C   54.0   12.0    877.0   148.0   
15          15    2021-22  32.0  LAC  NBA    C   35.0   10.0    538.0    94.0   
16          16    2021-22  32.0  MIL  NBA    C   19.0    2.0    339.0    54.0   
17          17     Career   NaN  NaN  NBA  NaN  903.0  697.0  24941.0  4549.0   
18          18        NaN   NaN  NaN  NaN  NaN    NaN    NaN      NaN     NaN   
19          19  7 seasons   NaN  OKC  NBA  NaN  524.0  413.0  15099.0  2558.0   
20          20  4 seasons   NaN  TOR  NBA  NaN  228.0  177.0   6300.0  1314.0   
21          21  2 seasons   NaN  LAC  NBA  NaN   76.0   49.0   1493.0   280.0   
22          22   1 season   NaN  ORL  NBA  NaN   56.0   56.0   1710.0   343.0   
23          23   1 season   NaN  MIL  NBA  NaN   19.0    2.0    339.0    54.0   

    ...     DRB     TRB    AST    STL     BLK     TOV      PF      PTS  \
0   ...   259.0   397.0   10.0   25.0    97.0    68.0   194.0    461.0   
1   ...   408.0   622.0   22.0   31.0   198.0    71.0   267.0    808.0   
2   ...   305.0   498.0   28.0   33.0   241.0    79.0   179.0    603.0   
3   ...   395.0   617.0   43.0   28.0   242.0   125.0   213.0   1055.0   
4   ...   485.0   709.0   85.0   39.0   219.0   123.0   233.0   1227.0   
5   ...   364.0   499.0   55.0   30.0   155.0    97.0   194.0    917.0   
6   ...   392.0   533.0   66.0   38.0   148.0   108.0   211.0    983.0   
7   ...   409.0   538.0   75.0   41.0   124.0   105.0   215.0   1173.0   
8   ...   282.0   382.0   60.0   34.0    91.0    66.0   139.0    846.0   
9   ...   127.0   156.0   15.0    7.0    33.0    39.0    76.0    327.0   
10  ...   400.0   478.0   62.0   29.0    99.0    90.0   209.0    959.0   
11  ...   445.0   601.0   99.0   29.0   103.0   114.0   211.0   1112.0   
12  ...   338.0   452.0   77.0   28.0    46.0   109.0   153.0    845.0   
13  ...   202.0   275.0   74.0    9.0    47.0    44.0    76.0    454.0   
14  ...   188.0   250.0   49.0   11.0    33.0    43.0   102.0    365.0   
15  ...   110.0   150.0   36.0    7.0    26.0    33.0    56.0    232.0   
16  ...    78.0   100.0   13.0    4.0     7.0    10.0    46.0    133.0   
17  ...  4590.0  6469.0  745.0  371.0  1752.0  1176.0  2457.0  10962.0   
18  ...     NaN     NaN    NaN    NaN     NaN     NaN     NaN      NaN   
19  ...  2608.0  3875.0  309.0  224.0  1300.0   671.0  1491.0   6054.0   
20  ...  1310.0  1687.0  253.0   93.0   281.0   352.0   649.0   3243.0   
21  ...   312.0   425.0  110.0   16.0    73.0    77.0   132.0    686.0   
22  ...   282.0   382.0   60.0   34.0    91.0    66.0   139.0    846.0   
23  ...    78.0   100.0   13.0    4.0     7.0    10.0    46.0    133.0   

    Unnamed: 30  Trp Dbl  
0           NaN      0.0  
1           NaN      0.0  
2           NaN      1.0  
3           NaN    

In [7]:
h5_data = pd.HDFStore('test.h5')

In [8]:
h5_data['obj1']=form_file

C:\Users\Hp\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3418: PerformanceWarning: 
your performance may suffer as PyTables will pickle object types that it cannot
map directly to c-types [inferred_type->mixed,key->block2_values] [items->Index(['Season', 'Tm', 'Lg', 'Pos'], dtype='object')]

  exec(code_obj, self.user_global_ns, self.user_ns)


In [9]:
h5_data['obj1']

Unnamed: 0     Season   Age   Tm   Lg  Pos      G     GS       MP      FG  \
0            0    2009-10  20.0  OKC  NBA   PF   73.0    0.0   1323.0   201.0   
1            1    2010-11  21.0  OKC  NBA   PF   82.0   44.0   2216.0   335.0   
2            2    2011-12  22.0  OKC  NBA   PF   66.0   66.0   1792.0   262.0   
3            3    2012-13  23.0  OKC  NBA   PF   80.0   80.0   2486.0   446.0   
4            4    2013-14  24.0  OKC  NBA   PF   81.0   81.0   2666.0   524.0   
5            5    2014-15  25.0  OKC  NBA   PF   64.0   64.0   2116.0   374.0   
6            6    2015-16  26.0  OKC  NBA   PF   78.0   78.0   2500.0   416.0   
7            7    2016-17  27.0  TOT  NBA   PF   79.0   79.0   2422.0   471.0   
8            8    2016-17  27.0  ORL  NBA   PF   56.0   56.0   1710.0   343.0   
9            9    2016-17  27.0  TOR  NBA   PF   23.0   23.0    712.0   128.0   
10          10    2017-18  28.0  TOR  NBA   PF   76.0   76.0   2093.0   379.0   
11          11    2018-19  29.0  TOR  NBA    C   74.0   51.0   2010.0   464.0   
12          12    2019-20  30.0  TOR  NBA    C   55.0   27.0   1485.0   343.0   
13          13    2020-21  31.0  LAC  NBA    C   41.0   39.0    955.0   186.0   
14          14    2021-22  32.0  TOT  NBA    C   54.0   12.0    877.0   148.0   
15          15    2021-22  32.0  LAC  NBA    C   35.0   10.0    538.0    94.0   
16          16    2021-22  32.0  MIL  NBA    C   19.0    2.0    339.0    54.0   
17          17     Career   NaN  NaN  NBA  NaN  903.0  697.0  24941.0  4549.0   
18          18        NaN   NaN  NaN  NaN  NaN    NaN    NaN      NaN     NaN   
19          19  7 seasons   NaN  OKC  NBA  NaN  524.0  413.0  15099.0  2558.0   
20          20  4 seasons   NaN  TOR  NBA  NaN  228.0  177.0   6300.0  1314.0   
21          21  2 seasons   NaN  LAC  NBA  NaN   76.0   49.0   1493.0   280.0   
22          22   1 season   NaN  ORL  NBA  NaN   56.0   56.0   1710.0   343.0   
23          23   1 season   NaN  MIL  NBA  NaN   19.0    2.0    339.0    54.0   

    ...     DRB     TRB    AST    STL     BLK     TOV      PF      PTS  \
0   ...   259.0   397.0   10.0   25.0    97.0    68.0   194.0    461.0   
1   ...   408.0   622.0   22.0   31.0   198.0    71.0   267.0    808.0   
2   ...   305.0   498.0   28.0   33.0   241.0    79.0   179.0    603.0   
3   ...   395.0   617.0   43.0   28.0   242.0   125.0   213.0   1055.0   
4   ...   485.0   709.0   85.0   39.0   219.0   123.0   233.0   1227.0   
5   ...   364.0   499.0   55.0   30.0   155.0    97.0   194.0    917.0   
6   ...   392.0   533.0   66.0   38.0   148.0   108.0   211.0    983.0   
7   ...   409.0   538.0   75.0   41.0   124.0   105.0   215.0   1173.0   
8   ...   282.0   382.0   60.0   34.0    91.0    66.0   139.0    846.0   
9   ...   127.0   156.0   15.0    7.0    33.0    39.0    76.0    327.0   
10  ...   400.0   478.0   62.0   29.0    99.0    90.0   209.0    959.0   
11  ...   445.0   601.0   99.0   29.0   103.0   114.0   211.0   1112.0   
12  ...   338.0   452.0   77.0   28.0    46.0   109.0   153.0    845.0   
13  ...   202.0   275.0   74.0    9.0    47.0    44.0    76.0    454.0   
14  ...   188.0   250.0   49.0   11.0    33.0    43.0   102.0    365.0   
15  ...   110.0   150.0   36.0    7.0    26.0    33.0    56.0    232.0   
16  ...    78.0   100.0   13.0    4.0     7.0    10.0    46.0    133.0   
17  ...  4590.0  6469.0  745.0  371.0  1752.0  1176.0  2457.0  10962.0   
18  ...     NaN     NaN    NaN    NaN     NaN     NaN     NaN      NaN   
19  ...  2608.0  3875.0  309.0  224.0  1300.0   671.0  1491.0   6054.0   
20  ...  1310.0  1687.0  253.0   93.0   281.0   352.0   649.0   3243.0   
21  ...   312.0   425.0  110.0   16.0    73.0    77.0   132.0    686.0   
22  ...   282.0   382.0   60.0   34.0    91.0    66.0   139.0    846.0   
23  ...    78.0   100.0   13.0    4.0     7.0    10.0    46.0    133.0   

    Unnamed: 30  Trp Dbl  
0           NaN      0.0  
1           NaN      0.0  
2           NaN      1.0  
3           NaN    

acroynm for such operation is SerDe <br>
ser-> object to bytecode (writing) <br>
De-> bytecode to object (reading) <br>

In [10]:
h5_data.put('obj2',form_file,format='table')

In [12]:
h5_data.select('obj2',where=['index >=10 and index <=15'])

Unnamed: 0   Season   Age   Tm   Lg Pos     G    GS      MP     FG  ...  \
10          10  2017-18  28.0  TOR  NBA  PF  76.0  76.0  2093.0  379.0  ...   
11          11  2018-19  29.0  TOR  NBA   C  74.0  51.0  2010.0  464.0  ...   
12          12  2019-20  30.0  TOR  NBA   C  55.0  27.0  1485.0  343.0  ...   
13          13  2020-21  31.0  LAC  NBA   C  41.0  39.0   955.0  186.0  ...   
14          14  2021-22  32.0  TOT  NBA   C  54.0  12.0   877.0  148.0  ...   
15          15  2021-22  32.0  LAC  NBA   C  35.0  10.0   538.0   94.0  ...   

      DRB    TRB   AST   STL    BLK    TOV     PF     PTS  Unnamed: 30  \
10  400.0  478.0  62.0  29.0   99.0   90.0  209.0   959.0          NaN   
11  445.0  601.0  99.0  29.0  103.0  114.0  211.0  1112.0          NaN   
12  338.0  452.0  77.0  28.0   46.0  109.0  153.0   845.0          NaN   
13  202.0  275.0  74.0   9.0   47.0   44.0   76.0   454.0          NaN   
14  188.0  250.0  49.0  11.0   33.0   43.0  102.0   365.0          NaN   
15  110.0  150.0  36.0   7.0   26.0   33.0   56.0   232.0          NaN   

    Trp Dbl  
10      0.0  
11      0.0  
12      0.0  
13      0.0  
14      0.0  
15      0.0  

[6 rows x 33 columns]

Here it is to be noted that, file will not be directly readbale in case of such different dat types so we have to use some other way to read it & supported by tools supported by it.<br>

These all are file systems, their application is important when we deal with huge amount of data. Bcz these file systems increases the SerDe operation useability with same computational power.Here the data remains the data is same the only change is the way it is storing the data.

Wherever optimizatioon is involved this format are very effective.


In [13]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

In [14]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [15]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [20]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [21]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [22]:
df.dtypes=='object'

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [23]:
#select only those dtypes where True values are there
df.dtypes[df.dtypes == 'object']

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
dtype: object

In [28]:
df.dtypes[df.dtypes == 'object'].index

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [25]:
df1 = df[df.dtypes[df.dtypes == 'object'].index]

In [27]:
df1

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

In [26]:
df1.describe()

Name   Sex  Ticket        Cabin Embarked
count                        891   891     891          204      889
unique                       891     2     681          147        3
top     Bystrom, Mrs. (Karolina)  male  347082  C23 C25 C27        S
freq                           1   577       7            4      644

In [29]:
#filter only those columns whose data types is only Float
df.dtypes=='float64'

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare            True
Cabin          False
Embarked       False
dtype: bool

In [31]:
df.dtypes[df.dtypes=='float64']

Age     float64
Fare    float64
dtype: object

In [32]:
df.dtypes[df.dtypes=='float64'].index

Index(['Age', 'Fare'], dtype='object')

In [33]:
#filtering the selected data
df[df.dtypes[df.dtypes=='float64'].index]

Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [34]:
df[["Age","Fare"]]

Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [35]:
df['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

In [36]:
df['Name'][5:10]

5                                     Moran, Mr. James
6                              McCarthy, Mr. Timothy J
7                       Palsson, Master. Gosta Leonard
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                  Nasser, Mrs. Nicholas (Adele Achem)
Name: Name, dtype: object

In [37]:
df['Name'][5:10:2]

5                       Moran, Mr. James
7         Palsson, Master. Gosta Leonard
9    Nasser, Mrs. Nicholas (Adele Achem)
Name: Name, dtype: object

In [39]:
df[['Ticket']][5:16]

Ticket
5      330877
6       17463
7      349909
8      347742
9      237736
10    PP 9549
11     113783
12  A/5. 2151
13     347082
14     350406
15     248706

In [40]:
sorted(df['Ticket'])

['110152',
 '110152',
 '110152',
 '110413',
 '110413',
 '110413',
 '110465',
 '110465',
 '110564',
 '110813',
 '111240',
 '111320',
 '111361',
 '111361',
 '111369',
 '111426',
 '111427',
 '111428',
 '112050',
 '112052',
 '112053',
 '112058',
 '112059',
 '112277',
 '112379',
 '113028',
 '113043',
 '113050',
 '113051',
 '113055',
 '113056',
 '113059',
 '113501',
 '113503',
 '113505',
 '113505',
 '113509',
 '113510',
 '113514',
 '113572',
 '113572',
 '113760',
 '113760',
 '113760',
 '113760',
 '113767',
 '113773',
 '113776',
 '113776',
 '113781',
 '113781',
 '113781',
 '113781',
 '113783',
 '113784',
 '113786',
 '113787',
 '113788',
 '113789',
 '113789',
 '113792',
 '113794',
 '113796',
 '113798',
 '113798',
 '113800',
 '113803',
 '113803',
 '113804',
 '113806',
 '113806',
 '113807',
 '11668',
 '11668',
 '11751',
 '11751',
 '11752',
 '11753',
 '11755',
 '11765',
 '11767',
 '11767',
 '11769',
 '11771',
 '11774',
 '11813',
 '11967',
 '11967',
 '12233',
 '12460',
 '12749',
 '12749',
 '13049'

In [41]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [42]:
for i in df.columns:
    print(df[i])

0        1
1        2
2        3
3        4
4        5
      ... 
886    887
887    888
888    889
889    890
890    891
Name: PassengerId, Length: 891, dtype: int64
0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64
0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64
0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, 

In [43]:
#adding a new column but note that data is not unique
df['sawan']="kumar"

In [44]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  sawan  
0        0         A/5 21171   7.2500   NaN        S  kumar  
1        0          PC 17599  71.2833   C85        C  kumar  
2        0  STON/O2. 3101282   7.9250   NaN        S  kumar  
3        0            113803  53.1000  C123        S  kumar  
4        0            373450   8.0500   NaN        S  kumar  
..     ...               ...      ...   ...      ...    ...  
886      0            211536  13.0000   NaN        S  kumar  
887      0            112053  30.0000   B42        S  kumar  
888      2        W./C. 6607  23.4500   NaN        S  kumar  
889      0            111369  30.0000  C148        C  kumar  
890      0            370376   7.7500   NaN        Q  kumar  

[891 rows x 13 columns]

In [45]:
#here we have added a new column whose va;lues are asssigned same as that of passenger id
df['sawan1']=df['PassengerId']

In [46]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  sawan  sawan1  
0        0         A/5 21171   7.2500   NaN        S  kumar       1  
1        0          PC 17599  71.2833   C85        C  kumar       2  
2        0  STON/O2. 3101282   7.9250   NaN        S  kumar       3  
3        0            113803  53.1000  C123        S  kumar       4  
4        0            373450   8.0500   NaN        S  kumar       5  
..     ...               ...      ...   ...      ...    ...     ...  
886      0            211536  13.0000   NaN        S  kumar     887  
887      0            112053  30.0000   B42        S  kumar     888  
888      2        W./C. 6607  23.4500   NaN        S  kumar     889  
889      0            111369  30.0000  C148        C  kumar     890  
890      0            370376   7.7500   NaN        Q  kumar     891  

[891 rows x 14 columns]

In [47]:
#Calling categorical data
#will give entire data information as series along with its length, categories  
pd.Categorical(df['Cabin'])

[NaN, 'C85', NaN, 'C123', NaN, ..., NaN, 'B42', NaN, 'C148', NaN]
Length: 891
Categories (147, object): ['A10', 'A14', 'A16', 'A19', ..., 'F38', 'F4', 'G6', 'T']

In [48]:
pd.Categorical(df['Sex'])

['male', 'female', 'female', 'female', 'male', ..., 'male', 'female', 'female', 'male', 'male']
Length: 891
Categories (2, object): ['female', 'male']

In [49]:
pd.Categorical(df['Pclass'])

[3, 1, 3, 1, 3, ..., 2, 1, 3, 1, 3]
Length: 891
Categories (3, int64): [1, 2, 3]

In [50]:
df['Pclass'].unique

<bound method Series.unique of 0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64>

In [54]:
df['Pclass'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Pclass, Length: 891, dtype: bool

In [53]:
df['Cabin'].isnull()

0       True
1      False
2       True
3      False
4       True
       ...  
886     True
887    False
888     True
889    False
890     True
Name: Cabin, Length: 891, dtype: bool

In [56]:
df['Cabin'][df['Cabin'].isnull() == True]

0      NaN
2      NaN
4      NaN
5      NaN
7      NaN
      ... 
884    NaN
885    NaN
886    NaN
888    NaN
890    NaN
Name: Cabin, Length: 687, dtype: object

In [57]:
#extarct indexes of null values
ser = df['Cabin'][df['Cabin'].isnull() == True]
ser

0      NaN
2      NaN
4      NaN
5      NaN
7      NaN
      ... 
884    NaN
885    NaN
886    NaN
888    NaN
890    NaN
Name: Cabin, Length: 687, dtype: object

In [58]:
ser.index

Int64Index([  0,   2,   4,   5,   7,   8,   9,  12,  13,  14,
            ...
            878, 880, 881, 882, 883, 884, 885, 886, 888, 890],
           dtype='int64', length=687)

In [59]:
#We are able to get dataset were we dont have a null value
df.iloc[ser.index]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
5      male   NaN      0      0            330877   8.4583   NaN        Q   
7      male   2.0      3      1            349909  21.0750   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S   
885  female  39.0      0      5            382652  29.1250   NaN        Q   
886    male  27.0      0      0            211536  13.0000   NaN        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     sawan  sawan1  
0    kumar       1  
2    kumar       3  
4    kumar       5  
5    kumar       6  
7    kumar       8  
..     ...     ...  
884  kumar     885  
885  kumar     886  
886  kumar     887  
888  kumar     889  
890  kumar     891  

[687 rows x 14 columns]

In [60]:
len(ser.index)

687

In [61]:
#extract a person name who has paid highest fare
#step-1
df['Fare']==max(df['Fare'])

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Fare, Length: 891, dtype: bool

In [62]:
# step-2 listing out highest fare as dataframe
df[df['Fare']==max(df['Fare'])]

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   

     sawan  sawan1  
258  kumar     259  
679  kumar     680  
737  kumar     738

In [63]:
#step-3 extracting name
df[df['Fare']==max(df['Fare'])]['Name']


258                      Ward, Miss. Anna
679    Cardeza, Mr. Thomas Drake Martinez
737                Lesurer, Mr. Gustave J
Name: Name, dtype: object

In [64]:
#combining two coulms and updating the next columns
#combine age & passengerId

df['new_col'] = df['PassengerId']  + df['Age']
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  sawan  sawan1  new_col  
0        0         A/5 21171   7.2500   NaN        S  kumar       1     23.0  
1        0          PC 17599  71.2833   C85        C  kumar       2     40.0  
2        0  STON/O2. 3101282   7.9250   NaN        S  kumar       3     29.0  
3        0            113803  53.1000  C123        S  kumar       4     39.0  
4        0            373450   8.0500   NaN        S  kumar       5     40.0  
..     ...               ...      ...   ...      ...    ...     ...      ...  
886      0            211536  13.0000   NaN        S  kumar     887    914.0  
887      0            112053  30.0000   B42        S  kumar     888    907.0  
888      2        W./C. 6607  23.4500   NaN        S  kumar     889      NaN  
889      0            111369  30.0000  C148        C  kumar     890    916.0  
890      0            370376   7.7500   NaN        Q  kumar     891    923.0  

[891 rows x 15 columns]

In [65]:
df5 = pd.read_csv('https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv',sep= '\t')
df5

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

In [67]:
df5.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

Some other links for data for practice
https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv
https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt
https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv
https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv
https://drive.google.com/file/d/1etGP85wJWsgqbsh3gX5oxc8dPKLMA1eT/view


## THANK YOU & KEEP LEARNING